contracts/src/v0.8/dev/transmission/entrypoint_implementation/contracts/Paymaster.sol

```diff
@@ -0,0 +1,118 @@
+
+  LinkTokenInterface public immutable LINK;
+  AggregatorV3Interface public immutable LINK_ETH_FEED;
+
+  struct Config {
+    uint32 stalenessSeconds;
+    int256 fallbackWeiPerUnitLink;
+  }
+  Config public s_config;
+
+  mapping(bytes32 => bool) userOpHashMapping;
+  mapping(address => uint256) subscriptions;
+
+  constructor(LinkTokenInterface linkToken, AggregatorV3Interface linkEthFeed) {
+    LINK = linkToken;
+    LINK_ETH_FEED = linkEthFeed;
+  }
+
+  function setConfig(uint32 stalenessSeconds, int256 fallbackWeiPerUnitLink) external {
+    s_config = Config({stalenessSeconds: stalenessSeconds, fallbackWeiPerUnitLink: fallbackWeiPerUnitLink});
+  }
+
+  function onTokenTransfer(address /* _sender */, uint256 _amount, bytes calldata _data) external {
+    if (msg.sender != address(LINK)) {
+      revert OnlyCallableFromLink();
+    }
+    if (_data.length != 32) {
+      revert InvalidCalldata();
+    }
+
+    address subscription = abi.decode(_data, (address));
+    subscriptions[subscription] += _amount;
+  }
+
+  function validatePaymasterUserOp(
+    UserOperation calldata userOp,
+    bytes32 userOpHash,
+    uint256 maxCost
+  ) external returns (bytes memory context, uint256 validationData) {
+    if (userOpHashMapping[userOpHash]) {
+      revert UserOperationAlreadyTried(userOpHash);
+    }
+
+    uint256 extraCostJuels = extractExtraCostJuels(userOp);
+    uint256 costJuels = getCostJuels(maxCost) + extraCostJuels;
+    if (subscriptions[userOp.sender] < costJuels) {
+      revert InsufficientFunds(costJuels, subscriptions[userOp.sender]);
+    }
+
+    userOpHashMapping[userOpHash] = true;
+    return (abi.encode(userOp.sender, extraCostJuels), _packValidationData(false, 0, 0)); // success
+  }
+
+  /// @dev Calculates any extra LINK cost for the user operation, based on the funding type passed to the
+  /// @dev paymaster.
+  /// TODO: add logic for subscription top-up.
+  function extractExtraCostJuels(UserOperation calldata userOp) internal returns (uint256 extraCost) {
+    if (userOp.paymasterAndData.length == 20) {
+      return 0; // no extra data, stop here
+    }
+
+    uint8 paymentType = uint8(userOp.paymasterAndData[20]);
+
+    // For direct funding, use top-up logic.
+    if (paymentType == uint8(SCALibrary.LinkPaymentType.DIRECT_FUNDING)) {
+      SCALibrary.DirectFundingData memory directFundingData = abi.decode(
+        userOp.paymasterAndData[21:],
+        (SCALibrary.DirectFundingData)
+      );
+      if (
+        directFundingData.topupThreshold != 0 &&
+        LINK.balanceOf(directFundingData.recipient) < directFundingData.topupThreshold
+      ) {
+        LINK.transfer(directFundingData.recipient, directFundingData.topupAmount);
+        extraCost = directFundingData.topupAmount;
+      }
+    }
+  }
+
+  /// @dev Deducts user subscription balance after execution.
+  function postOp(PostOpMode /* mode */, bytes calldata context, uint256 actualGasCost) external {
```

Comment:
Will this execute even if the user operation call reverts?

Reply:
Yes. The user will be billed for gas spent on a reverted end-user execution, and also for LINK transferred in the `validatePaymasterUserOp` portion. There are off-chain simulations done before a user operation is accepted to be executed on-chain. It is also up to the user to send valid user operations. A failing userOperation does not prevent LINK from being transferred, or from the user to be correctly billed.

Reply:
Ah actually I do see some refund logic in `EntryPoint.sol` which I guess is kind of odd, I would expect to see refunds in the Paymaster.

Reply:
`PostOpMode` gives three different cases: 
```
    enum PostOpMode {
        opSucceeded, // user op succeeded
        opReverted, // user op reverted. still has to pay for gas.
        postOpReverted //user op succeeded, but caused postOp to revert. Now it's a 2nd call, after user's op was deliberately reverted.
    }
```
However, I'm not sure this current implementation needs to take into account the `PostOpMode`, it seems that `actualGasCost` should correctly cover the true gas spent on the user operation regardless of success/failure executing the user operation of the gas limit given.

Reply:
Could be wrong here, and also haven't tested out billing to the extent I want to.